tower-http/src/propagate_header.rs

```diff
@@ -0,0 +1,87 @@
+use futures_util::ready;
+use http::{header::HeaderName, HeaderValue, Request, Response};
+use pin_project::pin_project;
+use std::future::Future;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+#[derive(Clone, Debug)]
+pub struct PropagateHeaderLayer {
+    header: HeaderName,
+}
+
+impl PropagateHeaderLayer {
+    pub fn new(header: HeaderName) -> Self {
+        Self { header }
+    }
+}
+
+impl<S> Layer<S> for PropagateHeaderLayer {
+    type Service = PropagateHeader<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        PropagateHeader {
+            inner,
+            header: self.header.clone(),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct PropagateHeader<S> {
```

Comment:
nit: would be nice if this had a constructor for building the service directly. it could also be good to add accessors for the inner service, and maybe also a `PropagateHeader::layer` function?

Reply:
Yes totally. Once a few PRs have been merged I'm gonna do a pass where I add inner service accessors and align the layers and services. I just got super tired of writing the same mostly boilerplate code over and over and wanted to initially focus more on the interesting part of the APIs.